### PR TITLE
Upsert

### PR DESCRIPTION
**Pull request by shulkaolka**

### CODE REVIEW

File: src/main/java/io/elastic/jdbc/ColumnNamesProvider.java [hunk 1]
```diff
@@ -1,113 +1,131 @@
...
-        if (configuration.get("tableName") == null || configuration.get("tableName").getAsString().isEmpty()) {
-            throw new RuntimeException("Table name is required");
+    } catch (SQLException e) {
+      throw new RuntimeException(e);
+    } finally {
+      if (rs != null) {
+        try {
+          rs.close();
+        } catch (SQLException e) {
+          logger.error("Failed to close result set {}", e.toString());
```
Comment (ShkarupaNick) [inline src/main/java/io/elastic/jdbc/ColumnNamesProvider.java]: You don`t need to call toString() from exception

File: src/main/java/io/elastic/jdbc/ColumnNamesProvider.java [hunk 2]
```diff
@@ -1,113 +1,131 @@
...
-                }
-            }
-            if (connection != null) {
-                try {
-                    connection.close();
-                } catch (SQLException e) {
-                    logger.error("Failed to close connection", e.toString());
-                }
-            }
+          logger.error("Failed to close connection {}", e.toString());
```
Comment (ShkarupaNick) [inline src/main/java/io/elastic/jdbc/ColumnNamesProvider.java]: You don`t need to call toString() from exception

File: src/main/java/io/elastic/jdbc/PrimaryColumnNamesProvider.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        .isEmpty()) {
+      throw new RuntimeException("Table name is required");
+    }
+    String tableName = configuration.getString("tableName");
+    JsonObject properties = Json.createObjectBuilder().build();
+    Connection connection = null;
+    ResultSet rs = null;
+    String schemaName = null;
+    Boolean isEmpty = true;
+    Boolean isOracle = (configuration.getString("dbEngine").equals("oracle")) ? true : false;
```
Comment (ShkarupaNick) [inline src/main/java/io/elastic/jdbc/PrimaryColumnNamesProvider.java]: wtf?  what sense of  ? true : false????

File: src/main/java/io/elastic/jdbc/PrimaryColumnNamesProvider.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    Boolean isMssql = (configuration.getString("dbEngine").equals("mssql")) ? true : false;
```
Comment (ShkarupaNick) [inline src/main/java/io/elastic/jdbc/PrimaryColumnNamesProvider.java]: the same  - true : false????

File: src/main/java/io/elastic/jdbc/PrimaryColumnNamesProvider.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+      }
+
+    } catch (SQLException e) {
+      throw new RuntimeException(e);
+    } finally {
+      if (rs != null) {
+        try {
+          rs.close();
+        } catch (SQLException e) {
+          logger.error("Failed to close result set", e.toString());
```
Comment (ShkarupaNick) [inline src/main/java/io/elastic/jdbc/PrimaryColumnNamesProvider.java]: logger.error("Failed to close result set {}", e);

File: src/main/java/io/elastic/jdbc/PrimaryColumnNamesProvider.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        }
+      }
+      if (connection != null) {
+        try {
+          connection.close();
+        } catch (SQLException e) {
+          logger.error("Failed to close connection", e.toString());
```
Comment (ShkarupaNick) [inline src/main/java/io/elastic/jdbc/PrimaryColumnNamesProvider.java]: logger.error("Failed to close connection {}", e);
where is exception throwing ?

File: src/main/java/io/elastic/jdbc/QueryBuilders/MSSQL.java [hunk 7]
```diff
@@ -1,27 +1,150 @@
...
+  public boolean executeRecordExists(Connection connection, JsonObject body) throws SQLException {
+    validateQuery();
+    String sql = "SELECT COUNT(*)" +
+        " FROM " + tableName +
+        " WHERE " + lookupField + " = ?";
+    PreparedStatement stmt = connection.prepareStatement(sql);
+    Utils.setStatementParam(stmt, 1, lookupField, body);
+    ResultSet rs = stmt.executeQuery();
+    rs.next();
+    return rs.getInt(1) > 0;
```
Comment (ShkarupaNick) [inline src/main/java/io/elastic/jdbc/QueryBuilders/MSSQL.java]: need stmt closing?

File: src/main/java/io/elastic/jdbc/QueryBuilders/MSSQL.java [hunk 8]
```diff
@@ -1,27 +1,150 @@
...
+    String sql = "INSERT INTO " + tableName +
+        " (" + keys.toString() + ")" +
+        " VALUES (" + values.toString() + ")";
+    PreparedStatement stmt = connection.prepareStatement(sql);
+    int i = 1;
+    for (Map.Entry<String, JsonValue> entry : body.entrySet()) {
+      Utils.setStatementParam(stmt, i, entry.getKey(), body);
+      i++;
+    }
+    stmt.execute();
```
Comment (ShkarupaNick) [inline src/main/java/io/elastic/jdbc/QueryBuilders/MSSQL.java]: need stmt closing?

File: src/main/java/io/elastic/jdbc/QueryBuilders/MySQL.java [hunk 9]
```diff
@@ -1,24 +1,146 @@
...
+        return stmt.executeQuery();
+    }
+
+    public ResultSet executeSelectTrigger(Connection connection, String sqlQuery) throws SQLException {
+        StringBuilder sql = new StringBuilder(sqlQuery);
+        PreparedStatement stmt = connection.prepareStatement(sql.toString());
+        if(pollingValue != null) {
+            stmt.setTimestamp(1, pollingValue);
+        }
+        return stmt.executeQuery();
```
Comment (ShkarupaNick) [inline src/main/java/io/elastic/jdbc/QueryBuilders/MySQL.java]: need stmt closing?

File: src/main/java/io/elastic/jdbc/QueryBuilders/MySQL.java [hunk 10]
```diff
@@ -1,24 +1,146 @@
...
         }
         sql.append(" ASC LIMIT ? OFFSET ?");
 
         PreparedStatement stmt = connection.prepareStatement(sql.toString());
-        stmt.setInt(1, countNumber);
-        stmt.setInt(2, skipNumber);
+        stmt.setTimestamp(1, pollingValue);
+        stmt.setInt(2, countNumber);
+        stmt.setInt(3, skipNumber);
         return stmt.executeQuery();
```
Comment (ShkarupaNick) [inline src/main/java/io/elastic/jdbc/QueryBuilders/MySQL.java]: need stmt closing?

File: src/main/java/io/elastic/jdbc/TableNameProvider.java [hunk 11]
```diff
@@ -1,84 +1,88 @@
...
-                }
-                result.addProperty(tableName, tableName);
-                isEmpty = false;
-            }
-            if (isEmpty) {
-                result.addProperty("", "no tables");
-            }
+      while (rs.next()) {
+        tableName = rs.getString("TABLE_NAME");
+        schemaName = rs.getString("TABLE_SCHEM");
```
Comment (ShkarupaNick) [inline src/main/java/io/elastic/jdbc/TableNameProvider.java]: it is normal "TABLE_SCHEM" ??? May be "TABLE_SCHEM**A**" ??

File: src/main/java/io/elastic/jdbc/Utils.java [hunk 12]
```diff
@@ -1,75 +1,307 @@
...
+      }
+      rs = md.getColumns(null, schemaName, tableName, "%");
+      while (rs.next()) {
+        String name = rs.getString("COLUMN_NAME").toLowerCase();
+        String type = detectColumnType(rs.getInt("DATA_TYPE"));
+        columnTypes.put(name, type);
+      }
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    } finally {
```
Comment (ShkarupaNick) [inline src/main/java/io/elastic/jdbc/Utils.java]: should we close rs in finally block?
May be we can move it above and remove catch block?

File: src/main/java/io/elastic/jdbc/ColumnNamesProvider.java (hunk 1 above)
Comment (shulkaolka) [reply src/main/java/io/elastic/jdbc/ColumnNamesProvider.java]: done

File: src/main/java/io/elastic/jdbc/PrimaryColumnNamesProvider.java (hunk 3 above)
Comment (shulkaolka) [reply src/main/java/io/elastic/jdbc/PrimaryColumnNamesProvider.java]: done

File: src/main/java/io/elastic/jdbc/PrimaryColumnNamesProvider.java (hunk 4 above)
Comment (shulkaolka) [reply src/main/java/io/elastic/jdbc/PrimaryColumnNamesProvider.java]: done

File: src/main/java/io/elastic/jdbc/PrimaryColumnNamesProvider.java (hunk 5 above)
Comment (shulkaolka) [reply src/main/java/io/elastic/jdbc/PrimaryColumnNamesProvider.java]: done

File: src/main/java/io/elastic/jdbc/PrimaryColumnNamesProvider.java (hunk 6 above)
Comment (shulkaolka) [reply src/main/java/io/elastic/jdbc/PrimaryColumnNamesProvider.java]: no need changes (was discussed)

File: src/main/java/io/elastic/jdbc/TableNameProvider.java (hunk 11 above)
Comment (shulkaolka) [reply src/main/java/io/elastic/jdbc/TableNameProvider.java]: no need changes (was discussed)

File: src/main/java/io/elastic/jdbc/Utils.java (hunk 12 above)
Comment (shulkaolka) [reply src/main/java/io/elastic/jdbc/Utils.java]: we should (was discussed)

File: src/main/java/io/elastic/jdbc/QueryBuilders/MSSQL.java (hunk 7 above)
Comment (shulkaolka) [reply src/main/java/io/elastic/jdbc/QueryBuilders/MSSQL.java]: done

File: src/main/java/io/elastic/jdbc/QueryBuilders/MySQL.java (hunk 10 above)
Comment (shulkaolka) [reply src/main/java/io/elastic/jdbc/QueryBuilders/MySQL.java]: done

File: src/main/java/io/elastic/jdbc/QueryBuilders/MSSQL.java (hunk 8 above)
Comment (shulkaolka) [reply src/main/java/io/elastic/jdbc/QueryBuilders/MSSQL.java]: done

File: src/main/java/io/elastic/jdbc/QueryBuilders/MySQL.java (hunk 9 above)
Comment (shulkaolka) [reply src/main/java/io/elastic/jdbc/QueryBuilders/MySQL.java]: done

**Pull request closed without merging**